`prefer-const` not triggered (regression from 2.4.0)

**What version of ESLint are you using?**
2.7.0

**What parser (default, Babel-ESLint, etc.) are you using?**
>  "babel-eslint": "~6.0.0"

**Please show your full configuration:**
```
{
  "parser": "babel-eslint",
  "extends": "airbnb",
  "rules": {
      "semi": [2, "never"],
      "arrow-body-style": 0
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
```
import React, { PropTypes } from 'react'

const innerStyle = {
  color: '#343434',
  color2: 'white',
}

const ToolBar = ({ style, draw }) => {
  let dynStyle = {
    ...innerStyle,
    ...style,
  }

  dynStyle.color = draw ? dynStyle.color2 : dynStyle.color

  return (
    <div style={dynStyle} />
  )
}

ToolBar.propTypes = {
  style: PropTypes.object,
  draw: PropTypes.bool,
}

export default ToolBar
```

**What did you expect to happen?**
`prefer-const` should trigger and ask me to change `let dynStyle` to `const dynStyle`

**What actually happened? Please include the actual, raw output from ESLint.**
ESLint don't trigger the `prefer-const` rule.

---

Note I tried these versions :
 - `~2.4.0` : **Works** 
 - `~2.5.0` : **Doesn't** work
 - `~2.6.0` : **Doesn't** work
 - `~2.7.0` : **Doesn't** work

---

Commits that are related to `prefer-const` (maybe I miss some) :
 - `2.6`: 9b73ffd Update: destructuring option of prefer-const rule (fixes #5594) (Toru Nagashima)
 - `2.5`: 9a22625 Fix: prefer-const false positive at non-blocked if (fixes #5610) (Toru Nagashima)
 - `2.5`: 87d74b2 Fix: prefer-const got to not change scopes (refs #5284) (Toru Nagashima)

Thank you for this issue.

Hm, I cannot reproduce this.
`prefer-const` is warning `dynStyle`.

```
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> cat test.js
import React, { PropTypes } from 'react'

const innerStyle = {
  color: '#343434',
  color2: 'white',
}

const ToolBar = ({ style, draw }) => {
  let dynStyle = {
    ...innerStyle,
    ...style,
  }

  dynStyle.color = draw ? dynStyle.color2 : dynStyle.color

  return (
    <div style={dynStyle} />
  )
}

ToolBar.propTypes = {
  style: PropTypes.object,
  draw: PropTypes.bool,
}

export default ToolBar

C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> eslint --version
v2.7.0
C:\Users\t-nagashima.AD\Documents\GitHub\eslint [master]> eslint test.js --rule prefer-const:2 --parser-options "{ecmaVersion:7,sourceType:'module',ecmaFeatures:{jsx:true,experimentalObjectRestSpread:true}}" --no-ignore --no-eslintrc

C:\Users\t-nagashima.AD\Documents\GitHub\eslint\test.js
  9:7  error  'dynStyle' is never reassigned, use 'const' instead  prefer-const

✖ 1 problem (1 error, 0 warnings)

--- a/lib/rules/prefer-const.js
+++ b/lib/rules/prefer-const.js
@@ -70,9 +70,14 @@ function canBecomeVariableDeclaration(identifier) {
  *
  * If the variable should not change to const, this function returns null.
  * - If the variable is reassigned.
- * - If the variable is never initialized and assigned.
+ * - If the variable is never initialized nor assigned.
  * - If the variable is initialized in a different scope from the declaration.
  * - If the unique assignment of the variable cannot change to a declaration.
+ *   e.g. `if (a) b = 1` / `return (b = 1)`
+ * - If the variable is declared in the global scope and `eslintUsed` is `true`.
+ *   `/*exported foo` directive comment makes such variables. This rule does not
+ *   warn such variables because this rule cannot distinguish whether the
+ *   exported variables are reassigned or not.
  *
  * @param {escope.Variable} variable - A variable to get.
  * @param {boolean} ignoreReadBeforeAssign -
@@ -82,7 +87,7 @@ function canBecomeVariableDeclaration(identifier) {
  *      Otherwise, null.
  */
 function getIdentifierIfShouldBeConst(variable, ignoreReadBeforeAssign) {
-    if (variable.eslintUsed) {
+    if (variable.eslintUsed && variable.scope.type === "global") {
         return null;
     }
 

--- a/tests/lib/rules/prefer-const.js
+++ b/tests/lib/rules/prefer-const.js
@@ -18,6 +18,12 @@ const rule = require("../../../lib/rules/prefer-const"),
 
 const ruleTester = new RuleTester();
 
+ruleTester.defineRule("use-x", context => ({
+    VariableDeclaration() {
+        context.markVariableAsUsed("x");
+    }
+}));
+
 ruleTester.run("prefer-const", rule, {
     valid: [
         { code: "var x = 0;" },
@@ -63,6 +69,7 @@ ruleTester.run("prefer-const", rule, {
             parserOptions: { ecmaVersion: 6 }
         },
         { code: "/*exported a*/ let a; function init() { a = foo(); }", parserOptions: { ecmaVersion: 6 } },
+        { code: "/*exported a*/ let a = 1", parserOptions: { ecmaVersion: 6 } },
         { code: "let a; if (true) a = 0; foo(a);", parserOptions: { ecmaVersion: 6 } },
 
         // The assignment is located in a different scope.
@@ -89,7 +96,9 @@ ruleTester.run("prefer-const", rule, {
             code: "let x; function foo() { bar(x); } x = 0;",
             options: [{ignoreReadBeforeAssign: true}],
             parserOptions: {ecmaVersion: 6}
-        }
+        },
+
+
     ],
     invalid: [
         {
@@ -336,6 +345,20 @@ ruleTester.run("prefer-const", rule, {
             output: "let x; function foo() { bar(x); } x = 0;",
             parserOptions: {ecmaVersion: 6},
             errors: [{ message: "'x' is never reassigned. Use 'const' instead.", type: "Identifier", column: 5}]
-        }
+        },
+
+        // https://github.com/eslint/eslint/issues/5837
+        {
+            code: "/*eslint use-x:error*/ let x = 1",
+            output: "/*eslint use-x:error*/ const x = 1",
+            parserOptions: {ecmaVersion: 6, ecmaFeatures: {globalReturn: true}},
+            errors: [{ message: "'x' is never reassigned. Use 'const' instead.", type: "Identifier"}]
+        },
+        {
+            code: "/*eslint use-x:error*/ { let x = 1 }",
+            output: "/*eslint use-x:error*/ { const x = 1 }",
+            parserOptions: {ecmaVersion: 6},
+            errors: [{ message: "'x' is never reassigned. Use 'const' instead.", type: "Identifier"}]
+        },
     ]
 });

